File: builder/java/ResourcePoolConfig.java
public class ResourcePoolConfig {
    /**
     * 假设 name 为必填项，其他非必填项
     */
    private final String name;
    private final int maxTotal;
    private final int maxIdle;
    private final int minIdle;

    /**
     * 如果私有化构造函数，那么它只能由内部类 Builder 调用
     */
    private ResourcePoolConfig(Builder builder) {
        this.name = builder.name;
        this.maxTotal = builder.maxTotal;
        this.maxIdle = builder.maxIdle;
        this.minIdle = builder.minIdle;
    }

    /**
     * 如果只提供 Getter 方法，那么对象在 build 出来了之后它就是只读的（不能再通过 Setter 方法修改了，事实上因此属性也被设置为 final 的）
     */
    public String getName() {
        return name;
    }

    public int getMaxTotal() {
        return maxTotal;
    }

    public int getMaxIdle() {
        return maxIdle;
    }

    public int getMinIdle() {
        return minIdle;
    }

    /**
     * Builder 可以设计成 ResourcePoolConfig 的内部类，也可以设计成独立的非内部类 ResourcePoolConfigBuilder
     */
    public static class Builder {
        private static final int DEFAULT_MAX_TOTAL = 8;
        private static final int DEFAULT_MAX_IDLE = 8;
        private static final int DEFAULT_MIN_IDLE = 0;

        private String name;
        /**
         * 非必填项，初始赋予默认值
         */
        private int maxTotal = DEFAULT_MAX_TOTAL;
        private int maxIdle = DEFAULT_MAX_IDLE;
        private int minIdle = DEFAULT_MIN_IDLE;

        /**
         * 整体的属性校验逻辑可以放到这里来做，包括必填项校验、依赖关系校验、约束条件校验等，因为它是对象构建中最后一个被调用的方法
         */
        public ResourcePoolConfig build() {
            if (name == null || name.isBlank()) {
                throw new IllegalArgumentException("name cannot be blank");
            }
            if (maxIdle > maxTotal) {
                throw new IllegalArgumentException("maxIdle cannot be greater than maxTotal");
            }
            if (minIdle > maxTotal || minIdle > maxIdle) {
                throw new IllegalArgumentException("minIdle cannot be greater than maxTotal or maxIdle");
            }
            // 校验通过后，返回构建的对象
            return new ResourcePoolConfig(this);
        }

        /**
         * Setter 方法，可使用属性名称作为方法名，或者 setName 作为方法名
         */
        public Builder name(String name) {
            // 这个必填项校验此处可不做，因为 build 中已经处理（build 中必须提供兜底担保，因为本 Setter 方法可能不会被调用）
            if (name == null || name.isBlank()) {
                throw new IllegalArgumentException("name cannot be blank");
            }
            this.name = name;
            return this;
        }

        public Builder maxTotal(int maxTotal) {
            if (maxTotal < 0) {
                throw new IllegalArgumentException("maxTotal cannot be less than zero");
            }
            this.maxTotal = maxTotal;
            return this;
        }

        public Builder maxIdle(int maxIdle) {
            if (maxIdle < 0) {
                throw new IllegalArgumentException("maxIdle cannot be less than zero");
            }
            this.maxIdle = maxIdle;
            return this;
        }

        public Builder minIdle(int minIdle) {
            if (minIdle < 0) {
                throw new IllegalArgumentException("minIdle cannot be less than zero");
            }
            this.minIdle = minIdle;
            return this;
        }
    }

    @Override
    public String toString() {
        return "ResourcePoolConfig{" +
                "name='" + name + '\'' +
                ", maxTotal=" + maxTotal +
                ", maxIdle=" + maxIdle +
                ", minIdle=" + minIdle +
                '}';
    }
}

File: builder/go/resource_pool_config.go
package builder

import (
	"errors"
	"fmt"
	"strings"
)

type ResourcePoolConfig struct {
	name     string
	maxTotal int
	maxIdle  int
	minIdle  int
}

func (c *ResourcePoolConfig) String() string {
	return fmt.Sprintf("ResourcePoolConfig{name='%s', maxTotal=%d, maxIdle=%d, minIdle=%d}",
		c.name, c.maxTotal, c.maxIdle, c.minIdle)
}

func (c *ResourcePoolConfig) Name() string {
	return c.name
}

func (c *ResourcePoolConfig) MaxTotal() int {
	return c.maxTotal
}

func (c *ResourcePoolConfig) MaxIdle() int {
	return c.maxIdle
}

func (c *ResourcePoolConfig) MinIdle() int {
	return c.minIdle
}

// ============================================================
// 参考了《Uber Go 语言编码规范》实现：https://github.com/uber-go/guide/blob/master/style.md#functional-options

type ResourcePoolConfigOption interface {
	apply(*ResourcePoolConfig) error
}

type maxTotalOption int

func (o maxTotalOption) apply(c *ResourcePoolConfig) error {
	maxTotal := int(o)
	if maxTotal < 0 {
		return errors.New("maxTotal cannot be less than zero")
	}
	c.maxTotal = maxTotal
	return nil
}

func WithMaxTotal(maxTotal int) ResourcePoolConfigOption {
	return maxTotalOption(maxTotal)
}

type maxIdleOption struct {
	maxIdle int
}

func (o maxIdleOption) apply(c *ResourcePoolConfig) error {
	maxIdle := o.maxIdle
	if maxIdle < 0 {
		return errors.New("maxIdle cannot be less than zero")
	}
	c.maxIdle = maxIdle
	return nil
}

func WithMaxIdle(maxIdle int) ResourcePoolConfigOption {
	return maxIdleOption{maxIdle}
}

type minIdleOption int

func (o minIdleOption) apply(c *ResourcePoolConfig) error {
	minIdle := int(o)
	if minIdle < 0 {
		return errors.New("minIdle cannot be less than zero")
	}
	c.minIdle = minIdle
	return nil
}

func WithMinIdle(minIdle int) ResourcePoolConfigOption {
	return minIdleOption(minIdle)
}

const (
	defaultMaxTotal = 8
	defaultMaxIdle  = 8
	defaultMinIdle  = 0
)

func NewResourcePoolConfig(name string, opts ...ResourcePoolConfigOption) (*ResourcePoolConfig, error) {
	// name 为必填属性，所以将其作为单独参数传入并且进行校验
	if strings.TrimSpace(name) == "" {
		return nil, errors.New("name cannot be blank")
	}

	// 先设置默认值
	conf := ResourcePoolConfig{
		name:     name,
		maxTotal: defaultMaxTotal,
		maxIdle:  defaultMaxIdle,
		minIdle:  defaultMinIdle,
	}

	// 应用传入的所有配置选项
	for _, o := range opts {
		if err := o.apply(&conf); err != nil {
			return nil, err
		}
	}

	// 依赖关系校验、约束条件校验等
	if conf.maxIdle > conf.maxTotal {
		return nil, errors.New("maxIdle cannot be greater than maxTotal")
	}
	if conf.minIdle > conf.maxTotal || conf.minIdle > conf.maxIdle {
		return nil, errors.New("minIdle cannot be greater than maxTotal or maxIdle")
	}
	return &conf, nil
}
